docs/examples/image_morphology/image_morphology.jl

```diff
@@ -0,0 +1,132 @@
+# ---
+# cover:  assets/morphology.gif
+# title: Morphological Operations
+# author: Ashwani Rathee
+# date: 2021-7-2
+# ---
+
+# Morphological image processing is a area of computer vision which pursue tophat
+# remove imperfections from binary and grayscale images. ImageMorphology.jl
+# holds a collection of non-linear operations related to the shape or
+# morphology of features in an image.
+
+# In this demonstration, we cover the 6 most common morphological operations:
+# - Erosion
+# - Dilation
+# - Opening
+# - Closing
+# - Tophat
+# - Bothat 
+# - Morphological Gradient
+# - Morphological Laplace
+
+# Let's import an image 
+
+using Images, ImageView, ImageMorphology, TestImages, MosaicViews
+
+img = testimage("morphology_test_512")
+
+# This is an RGB image, which we would need to be convert to either grayscale
+# binary image as morphology functions only work on grayscale or binary images.
+
+# ## Erosion
+
+# Erode performs perform a min-filter over nearest-neighbors. The default 
+# is 8-connectivity in 2d, 27-connectivity in 3d, etc. You can specify 
+# the list of dimensions that you want to include in the connectivity,
+#  e.g., `region = [1,2]` would exclude the third dimension from filtering.
+
+img_erode = Gray.(Gray.(img) .< 0.8); # keeps white objects white
+img_erosion1 = erode(img_erode)
+img_erosion2 = erode(erode(img_erode))
+img_erosion3 = erode(erode(erode(img_erode)))
+mosaicview(img_erode, img_erosion1, img_erosion2, img_erosion3; nrow = 1)
+
+# ## Dilation
+
+# Dilate performs a max-filter over nearest-neighbors. The default is 
+# 8-connectivity in 2d, 27-connectivity in 3d, etc. You can specify the
+# list of dimensions that you want to include in the connectivity,
+#  e.g., `region = [1,2]` would exclude the third dimension from filtering.
+
+img_dilate = Gray.(1 * Gray.(img) .> 0.9);
```

Comment:
Head up:

- mixed-precision computation is slow (especially for FixedPointNumbers types like `N0f8`)
- mix broadcasting with some dots and some non-dots is slow: `1 * Gray.(img)` forces Julia to generate a full array as the intermediate result. (Notice the memory usage)


```julia
julia> img = testimage("morphology_test_512");

julia> mixed_f(img) = Gray.(1 * Gray.(img) .> 0.9);

julia> broadcast_f(img) = @. Gray(1 * Gray(img) > 0.9);

julia> manual_broadcast_f(img) = Gray.(1 .* Gray.(img) .> 0.9);

julia> @btime mixed_f($img);
  755.592 μs (6 allocations: 1.50 MiB)

julia> @btime broadcast_f($img);
  1.089 ms (2 allocations: 256.11 KiB)

julia> @btime manual_broadcast_f($img);
  1.088 ms (2 allocations: 256.11 KiB)

julia> img = float.(testimage("morphology_test_512"));

julia> @btime mixed_f($img);
  621.169 μs (6 allocations: 2.25 MiB)

julia> @btime broadcast_f($img);
  382.919 μs (2 allocations: 256.11 KiB)

julia> @btime manual_broadcast_f($img);
  382.830 μs (2 allocations: 256.11 KiB)
```